Cookbook/Proposals/PropertyWrappers.md

```diff
@@ -0,0 +1,155 @@
+
+The simplest example of a property wrapper would be if we make an `ISO8601Date` property wrapper. This only requires to add a `@propertyWrapper` annotation to this type. It already implements decoding for dates so we don't need to change that. What changes is only how we declare properties of this type.
+
+Before:
+
+```swift
+struct Appointment: Decodable {
+  let time: Date
+	
+  init(from decoder: Decoder) throws {
+    let values = try decoder.container(keyedBy: CodingKeys.self)
+    time = try values.decode(ISO8601Date.self, forKey: .time)
+  }
+}
+```
+
+After:
+
+```swift
+struct Appointment: Decodable {
+  @ISO8601Date
+  let time: Date
+}
+```
+
+Reference: https://speakerdeck.com/alisoftware/and-thats-a-wrap?slide=25
+
+Another example is pruned/strict array decoding. The pruned strategy means that any array element that fails decoding will be discarded whereas with the strict strategy the whole decoding will fail on any failure in decoding of any array element (default behaviour). To implement prune strategy we use `FailableDecodable` that decodes its value with `value = try? container.decode(T.self)` or `decodeArraySafely` method. With a property wrapper we wouldn't need to to call this method manually or change the type of the property. We can also implement more strategies, like non-empty, using the same property wrapper.
+
+Before:
+
+```swift
+struct DemographicsListDTO: Decodable {
+  let demographics: [DemographicDTO]
+	
+  init(from decoder: Decoder) throws {
+    let values = try decoder.container(keyedBy: CodingKeys.self)
+    let demographics = try values.decodeArraySafely([DemographicDTO].self, forKey: .demographics)
+  }
+}
+```
+
+After:
+
+```swift
+struct DemographicsListDTO: Decodable {
+  @ArrayDecodable<Prune, DemographicDTO>
+  let demographics: [DemographicDTO]
+}
+```
+
+Providing defaults for absent values has slightly tricky ergonomics, as this approach does not allow to pass arbitrary parameters to property wrapper initialisers when they are being decoded (so `@Default(value: true)` is possible programmatically but won't affect decoding). So instead of using values we use types which would correspond to common defaults, i.e. `True`, `False`, `Empty` etc:
+
+```swift
+struct Product: Codable {
+  var name: String
+  
+  @DefaultDecodable<Empty>
+  var description: String
+  
+  @DefaultDecodable<True>
+  var isAvailable: Bool
+}
+```
+
+Reference: https://github.com/gonzalezreal/DefaultCodable
+
+Nested keys are a bit more tricky, but still possible:
+
+```swift
+struct InboxMessage: Decodable {
+  @NestedDecodable<Bool, IsReadKeys>
+  var isRead: Bool // { "status": { "is_read": true } }
+  
+  enum CodingKeys: String, CodingKey {
+    case isRead = "status"
+  }
+  
+  enum IsReadKeys: String, CodingKey, CaseIterable {
+    case isRead = "is_read"
+  }
```

Comment:
Everything else is a lovely improvement! But I am not too sure about property wrapper for nested types, it does not look very intuitive and there is some boiler plate code involved.

Reply:
Can you elaborate what is not intuitive and what boilerplate is needed? The only thing that is needed is to define CodingKeys in a straightforward way, and only to do that for the nested key. Comparing with a boilerplate we otherwise have to write to decode it manually its much less. There are also alternative approaches to this case https://twitter.com/ilyapuchka/status/1228062000682610690?s=20

Reply:
For simple cases though we already have a solution (`decodeNested`) that we can keep using.